### ISSUE TITLE
"Allow the users to reset their password" config is redundant

### ISSUE THREAD

**Issue by Piedone** [OrchardCMS/OrchardCore#6205]
When you enable the Users Reset Password feature then users won't actually be able to reset passwords until the "Allow the users to reset their password" config is checked. I don't see why you'd ever want to have the feature enabled while disabling it; if you want to disable it just disable the feature, that's why it's a separate feature.

If somebody can make the case for the config then I'll accept it but I'd still make it the default to be enabled. Just as with https://github.com/OrchardCMS/OrchardCore/issues/6105 this is a counter-intuitive necessary second step: You want password reset to be available, so you enable the feature. Do you have password reset now? You don't, because you also have to enable it under the settings.

**Comment by JoshLefebvre** [comment]
I am in need of this. Just looked up all instance of ResetPasswordSettings AllowResetPassword and there are several lines in the ResetPasswordController.cs as follows:  
```csharp
            if (!(await _siteService.GetSiteSettingsAsync()).As<ResetPasswordSettings>().AllowResetPassword)
            {
                return NotFound();
            }
```

Looking at this controller I see the following attribute:
```csharp
    [Feature("OrchardCore.Users.ResetPassword")]
    public class ResetPasswordController : Controller
    {
```

So I am assuming I can just remove these additional checks ?? Remove the references from the drivers and the views as well?

**Comment by JoshLefebvre** [comment]
One issue that I just found in Login.cshtml:
```csharp
@{
    var allowResetPassword = (await SiteService.GetSiteSettingsAsync()).As<ResetPasswordSettings>().AllowResetPassword;
}
...
    @if (allowResetPassword)
     {
           <a asp-controller="ResetPassword" asp-action="ForgotPassword" class="label-link">@T["Forgot your password?"]</a>
     }
...

Perhaps we can check enabled features instead?

**Comment by Piedone** [comment]
I'm confused, what would you like to achieve exactly? If we get rid of the settings then references to that should be removed, yes, but not the `Feature` attribute form the controller.

**Comment by JoshLefebvre** [comment]
Yes sorry I should have been more clear. I meant keep the feature attribute, which will only allow users to navigate to these actions if the feature is enabled. The following line is redundant at this point so remove it:
```await _siteService.GetSiteSettingsAsync()).As<ResetPasswordSettings>().AllowResetPassword```

**Comment by Piedone** [comment]
Yeah, that's good. However, we haven't yet triaged this issue so there's no decision on whether to change it according to my suggestion.

If you just want to make things easier in your own app because e.g. you run a SaaS then I suggest configuring the checkbox to be ticked by default when the feature is enabled. You can do this e.g. from an `IFeatureEventHandler`.

**Comment by sebastienros** [comment]
Is there a workflow task also in the feature? That could explain.
Also maybe worth checking the history of the feature, it's possible that we talked about it and decided to create a setting on purpose. Also check if the users who can disable this settings can actually disable the feature.

**Comment by Piedone** [comment]
The feature-settings permission I thought about but that still sounds like a solution to a non-existing problem to me (do you really need one person to make it possible to configure password reset availability, and somebody else to be able to turn it on or off?), and if it exists an "on" default would be better. However, there's no separate permission for password reset configuration (only `ManageUsers` for everything).

**Comment by sebastienros** [comment]
Then that's a separate permission than enabling features. Imagine a multi tenant solution that has this feature enabled by default but customers can decide to allow reset password or not. Plus the other argument I mentioned where other things could be part of the feature. I don't see why we should remove the setting.

**Comment by Piedone** [comment]
This is very hard for me to imagine :). Anyway, have default as on?

**Comment by sebastienros** [comment]
On by default is totally fine. Seems like the correct fix to me.

**Comment by JoshLefebvre** [comment]
This is exactly what I did. I just added true as the default value in the ResetPasswordSettings as a quick fix in my code:
```csharp
namespace OrchardCore.Users.Models
{
    public class ResetPasswordSettings
    {
        public bool AllowResetPassword { get; set; } = true;
        public bool UseSiteTheme { get; set; }
    }
}
```

**Comment by deanmarcussen** [comment]
The issue here, is that setting `AllowResetPassword` to true will mean that the view will always show the reset password section, whether the feature is enabled or not. This is because `.As<ResetPasswordSettings>()` will create a `new T`, which would then be defaulted to true.

I agree the setting feels redundant, but this is why the setting is done how it is.

So what would need to happen to avoid having to have the setting, is the View itself would need to be zoned, such that, drivers / shape providers, could add the reset password section to a zone in the View, if the feature is enabled.

**Comment by deanmarcussen** [comment]
There is a further issue here, related to many of these _Features_ that also require a second setting.

When the setting has been set true, and the feature it then disabled, the setting remains active.

So it becomes very counterintuitive, because you have to then reenable the feature, to disabled the settings.

For example in this screen shot, the login screen still shows the _Reset Password Feature_ even though I have disabled the feature.

![Screenshot 2020-07-05 at 12 16 30](https://user-images.githubusercontent.com/13782679/86531490-b26df800-beb9-11ea-9316-80e55bad4018.png)

This affects many of the user features that have a setting to go with them.

In the screenshot the _Register new user_ feature is also inactive, but because the setting is set to true, but in both cases the links are no longer active, because the feature has been disabled.

Two possible approaches to resolve
- Automatically turn the setting of/on, using the Feature activated / deactivated events (so hack on top of a hack)
- Zone the login screen so drivers can provide shapes for these features when the feature is active.

**Comment by Piedone** [comment]
Ah, that's a can of worms.

I think yes, have zones would be not just the proper solution to these issues but also a way to make the login screen better extensible in the long run too.
